src/secret.ts

```diff
@@ -0,0 +1,221 @@
+import { aws_secretsmanager } from "aws-cdk-lib";
+import type AWS from "aws-sdk";
+import { Construct } from "constructs";
+import { NativePreWarmContext, PrewarmClients } from "./function-prewarm";
+import { makeIntegration } from "./integration";
+
+export interface SecretProps<SecretData = any>
+  extends aws_secretsmanager.SecretProps {
+  /**
+   * When true, the {@link SecretData} will be serialized to/from JSON.
+   *
+   * @default true
+   */
+  storeAsJson?: SecretData extends string ? false : boolean;
+}
+
+/**
+ * @returns `true` of {@link a} is a {@link Secret}, otherwise `false`.
+ */
+export function isSecret(a: any): a is Secret {
+  return a?.FunctionlessType === Secret.FunctionlessType;
+}
+
+/**
+ * Securely stores secret data such as credentials and API keys in an AWS Secret.
+ *
+ * @see https://aws.amazon.com/secrets-manager/
+ * @see {@link aws_secretsmanager.Secret}
+ */
+export class Secret<SecretData = any> {
+  public static readonly FunctionlessType = "Secret";
+  public readonly FunctionlessType = Secret.FunctionlessType;
+
+  readonly resource: aws_secretsmanager.ISecret;
+  readonly storeAsJson: boolean;
+
+  /**
+   * An Integration that stores a new value in the Secret.
+   *
+   * There are three ways to store secrets:
+   * 1. Recommended - passing the strongly typed `SecretValue`
+   *
+   * For the best experience, we recommend strongly typing secret data and serializing to/from JSON.
+   *
+   * ```ts
+   * interface UserPass {
+   *   username: string;
+   *   password: string;
+   * }
+   *
+   * const secret = new Secret<UserPass>(scope, id);
+   *
+   * await secret.putSecretValue({
+   *   SecretValue: {
+   *     username: "sam"
+   *     password: "my little pony"
+   *   }
+   * });
+   * ```
+   */
+  public readonly putSecretValue;
+
+  public readonly getJsonValue;
+
+  /**
+   * Get the
+   */
+  public readonly getValue;
+
+  constructor(resource: aws_secretsmanager.ISecret);
+  constructor(scope: Construct, id: string, props?: SecretProps);
+  constructor(
+    ...args:
+      | [aws_secretsmanager.ISecret]
+      | [scope: Construct, id: string, props?: SecretProps]
+  ) {
+    if (args.length === 1) {
+      this.resource = args[0];
+      this.storeAsJson = true;
+    } else {
+      const [scope, id, props] = args;
+      this.resource = new aws_secretsmanager.Secret(scope, id, props);
+      this.storeAsJson = props?.storeAsJson ?? true;
+    }
+
+    const storeAsJson = this.storeAsJson;
```

Comment:
I don't think we can assume it's always json. Maybe we a JsonSecret class?

Reply:
but isn't `this.storeAsJson` always accurate?

I was wondering about a separate class too. If we can simplify the contract I'd say make a separate class. If it reduces the risk of calling `getJson` when it will fail too,

Reply:
It's not the same. If the Secret is just an API key, e.g. `ABCD1234567`, that's not valid JSON. It would have to be `"ABCD1234567"`. There's a difference between a JSON-encoded string and arbitrary text.